### DISCUSSION TITLE
Very interesting project, how does it work?

### DISCUSSION THREAD

**Discussion by GavinRay97** [sfu-db/connector-x#356] [Q&A]
Hello, this project is super cool! Especially what looks like federated querying feature.

I took a look at the code, and from what I can understand it looks like:
- Apache Calcite `JdbcSchema`'s are created using JVM <-> Rust bridge for the datasource
- Calcite is used to optimize the query
- The schema of the query is used to create Arrow schema
- The query is sent over database's binary transport, and results streamed as Arrow `RecordBatch` objects

Is this correct?
I would love to understand more about the approach, it seems clever🙂

**Comment by wangxiaoying** [comment]
Hi @GavinRay97 , thank you for your interest in connectorx! You are basically correct with the workflow. In general, for federated queries we:
1. rewrite the SQL query into `n` queries
2. run the first `n-1` queries concurrently and get their results in arrow record batch
3. run the last query on the collected arrow record batches by joining them to get the final result (also in arrow record batch)

As you mentioned, we use Calcite in step 1 to optimize (push down some operators) and split the query into multiple ones (each run on a single data source). Since our main code is written in rust, we use `j4rs` to bridge rust and java through starting a JVM environment in rust and [invoke the compiled jar file to get the rewritten queries](https://github.com/sfu-db/connector-x/blob/main/connectorx/src/fed_dispatcher.rs#L103). Since Calcite needs to connect to the database to get the metadata (column names and types) in order to validate the SQL query, we need to input the database connection informations (in JDBC) through [`dataSource`s](https://github.com/sfu-db/connector-x/blob/main/connectorx/src/fed_dispatcher.rs#L56). The source code of the jar file is [here](https://github.com/sfu-db/federated-query) with a simple example of the input and output in the readme.

For step 2, we directly fetch the data from each datasource into arrow format through rust client drivers (note that unlike in Calcite which use JDBC to connect to sources, we use native rust drivers here), leveraging our optimizations in connectorx (Feel free to look at our [technical report](https://github.com/sfu-db/connector-x/blob/a69f8a388db4a9e9b9b51928aa0c3aa04bd6fd4e/assets/Technical_Report__ConnectorX.pdf) if you are interested in this part). For this part, the schema is actually refetched in rust for each datasource's implementation of the [`fetch_metadata` interface](https://github.com/sfu-db/connector-x/blob/main/connectorx/src/sources/mod.rs#L41). We define the type mapping relations in these [transport files](https://github.com/sfu-db/connector-x/tree/main/connectorx/src/transports).

For step 3, we directly leverage [`datafusion`](https://github.com/apache/arrow-datafusion) to run SQL on arrow record batches. Basically we treat the arrow result from each datasource a single table and name them as the source's name. Like `db1` and `db2` in `q_local` in the [simple example](https://github.com/sfu-db/federated-query#federated-query).

**Reply by GavinRay97** [reply]
Thank you!
